### ISSUE TITLE
capstone-rs backend?

### ISSUE THREAD

**Issue by iximeow** [endgameinc/xori#21]
I noticed yesterday that much of [Xori](https://github.com/endgameinc/xori/blob/master/src/arch/x86/prefixx86.rs#L149)'s instruction decoding maps pretty directly to chunks of [Capstone](https://github.com/aquynh/capstone/blob/master/arch/X86/X86DisassemblerDecoder.c#L488)'s decoding logic. Since that appears to be the case for so much of xori's decoder, does it make sense to use capstone-rs directly?

I noticed in your slides that you mention fixing a few capstone bugs in the process of rewriting, so maybe those fixes could be upstreamed? If the motivator for decoding in rust is avoiding memory bugs another thought might be to see about parcelling out Xori's x86 decoder and seeing about replacing the capstone decoder upstream?

Barring any of that, it would be good to know that Xori's x86 decoding very closely derives from Capstone's implementation, much like [Capstone does in indicating its LLVM origins](https://github.com/aquynh/capstone/blob/master/arch/X86/X86DisassemblerDecoder.c#L3), especially if bugs were found in Xori's decoder to know that they should be upstreamed - even the disassembly mapping tables are the same as Capstone's [and those do change from time to time](https://github.com/aquynh/capstone/commit/532bc3f328f68e7f7a330a6adcc5b214104b53d1)!

(a nice part of capstone-rs as a backend means the bar for supporting any other Capstone-backed arch is a lot easier!)

**Comment by securedorg** [comment]
Yeah the idea was to convert the logic only using rust since that was a requirement at work. I needed to verify all of the logic by hand to make sure I had parity with other disassemblers I was testing. There were some things in the capstone lib that were extra unused code that was pulled in from llvm libs (more in the printing).

**Comment by danghvu** [comment]
@securedorg Don't you need to credit or attribute somewhere @aquynh / capstone-engine / llvm  if you re-use/adapt their code ?

**Comment by rseymour** [comment]
I think we'd rather put more documentation in the README and in the code to give proper credit (to capstone, LLVM, etc) than just link back to the slides.  This is a long holiday weekend so I may not get to it until Tuesday, but ping me if you need it before then.  We will get this right as many iterations as it takes. Let me know if there is specific links, wording you'd like etc otherwise I'll look it up and do my best.

As per the original ticket, I'm not against adding capstone-rs as a xori backend. In addition to instant extra arch support, it would allow for better testing and parity between the xori engine and capstone. 

Our current roadmap (aka what Amanda and I can do in our free time) doesn't include adding swappable disassembly engines. No matter what we would always want to maintain xori's 'pure rust' engine.  I would love to help do the design work such that we could do that in conjunction with some folks w/ capstone experience, but we don't have time ourselves in the foreseeable future to code it up.  Also although I don't love the idea, forking xori w/ capstone, xoristone? capxori?  could conceivably be easier than making the disassembler a configuration option (or cargo build feature).

**Comment by rseymour** [comment]
#22 is a start at adding some proper acknowledgements.
